File: search_insert.rb
# return the target index, else return right

require 'pry'

def search_insert(nums, target)
  return nums.length if target > nums[nums.length - 1]
  return 0           if target < nums[0]

  binary_search(nums, target)
end

def binary_search(array, target)
  left = 0
  right = array.length - 1

  while left + 1 < right
    mid = left + (right - left) / 2
    if array[mid] == target
      return mid
    elsif array[mid] < target
      left = mid
    else
      right = mid
    end
  end

  return left if array[left] == target
  return right if array[right] == target

  right
end

puts search_insert([1,3,5,6], 5) #2
puts search_insert([1,3,5,6], 2) #1
puts search_insert([1,3,5,6], 7) #4
puts search_insert([1,3,5,6], 0) #0

File: binary_search/search_insert.js
// Given a sorted array and a target value,
// return the index if the target is found. If not,
// return the index where it would be if it were inserted in order.

// You may assume no duplicates in the array.

// Here are few examples.
// [1,3,5,6], 5 → 2
// [1,3,5,6], 2 → 1
// [1,3,5,6], 7 → 4
// [1,3,5,6], 0 → 0

// binary search
// return index when found
// else if not found, return mid

function searchInsert(nums, target) {
  if (target < nums[0])               return 0;
  if (target > nums[nums.length - 1]) return nums.length;
  return binarySearch(nums, target, 0, nums.length - 1);
}

function binarySearch(nums, target, low, high) {
  var mid = low + Math.floor((high - low) / 2);

  while (low + 1 < high) {
    if (nums[mid] === target) {
      return mid;
    } else if (nums[mid] > target) {
      return binarySearch(nums, target, low, mid);
    } else {
      return binarySearch(nums, target, mid, high);
    }
  }

  if (nums[low]  === target) return low;
  if (nums[high] === target) return high;
  return mid + 1;
}

console.log(searchInsert([1,3,5], 1));
